src/graph/backend/elyzor/compiler_loader.hpp

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+ * Copyright 2021-2024 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+#ifndef ELYZOR_GRAPH_COMPILER_LOADER_H
+#define ELYZOR_GRAPH_COMPILER_LOADER_H
+
+#include <iostream>
+#include "common/verbose.hpp"
+#include "include/dnnl_graph_compiler.h"
+
+namespace dnnl {
+namespace impl {
+namespace graph {
+namespace elyzor {
+
+typedef dnnl_status_t (*dnnl_graph_compiler_create_t)(
+        const struct dnnl_graph_compiler_context *ctx,
+        const struct dnnl_graph_compiler **gc);
+
+typedef void (*dnnl_graph_compiler_destroy_t)(
+        const struct dnnl_graph_compiler *gc);
+
+typedef dnnl_status_t (*dnnl_graph_compiler_compile_t)(
+        const struct dnnl_graph_compiler *gc, const char *graph_json,
+        const struct dnnl_graph_compiler_executable **exe);
+
+typedef void (*dnnl_graph_compiler_destroy_executable_t)(
+        const struct dnnl_graph_compiler *gc,
+        const struct dnnl_graph_compiler_executable *exe);
+
+typedef dnnl_status_t (*dnnl_graph_compiler_execute_t)(
+        const struct dnnl_graph_compiler *gc,
+        const struct dnnl_graph_compiler_executable *exe,
+        dnnl_graph_compiler_tensor *inputs,
+        dnnl_graph_compiler_tensor *outputs);
+
+struct dnnl_graph_compiler_vtable {
+    dnnl_graph_compiler_create_t dnnl_graph_compiler_create;
+    dnnl_graph_compiler_destroy_t dnnl_graph_compiler_destroy;
+    dnnl_graph_compiler_compile_t dnnl_graph_compiler_compile;
+    dnnl_graph_compiler_destroy_executable_t
+            dnnl_graph_compiler_destroy_executable;
+    dnnl_graph_compiler_execute_t dnnl_graph_compiler_execute;
+};
+
+class graph_compiler_loader {
+public:
+    static const dnnl_graph_compiler_vtable &get_vtable() {
+        static graph_compiler_loader ins;
+        return ins.vtable_;
+    }
+    ~graph_compiler_loader();
+
+private:
+    graph_compiler_loader();
+    graph_compiler_loader(graph_compiler_loader &) = delete;
+    graph_compiler_loader(graph_compiler_loader &&) = delete;
+    graph_compiler_loader &operator=(const graph_compiler_loader &) = delete;
+    graph_compiler_loader &operator=(graph_compiler_loader &&) = delete;
+
```

Comment:
_Originally posted by @AndreyPavlenko in https://github.com/intel-ai/oneDNN/pull/2#discussion_r1591441122_

Instead of using mutex, I recommend to use either a singleton or shared_ptr.
Here is a simple singleton-based implementation:
<details>

```c++
// graph_compiler.hpp

struct dnnl_graph_compiler {
    const dnnl_graph_compiler_get_api_version_t
            dnnl_graph_compiler_get_api_version;
    const dnnl_graph_compiler_create_t dnnl_graph_compiler_create;
    const dnnl_graph_compiler_destroy_t dnnl_graph_compiler_destroy;
    const dnnl_graph_compiler_compile_t dnnl_graph_compiler_compile;
    const dnnl_graph_compiler_destroy_executable_t
            dnnl_graph_compiler_destroy_executable;
    const dnnl_graph_compiler_execute_t dnnl_graph_compiler_execute;

    static const dnnl_graph_compiler_vtable &get() {
        void *dl = dlopen(DNNL_GC_LIB_NAME, RTLD_LAZY);
        // The singleton
        static dnnl_graph_compiler_vtable instance(
                reinterpret_cast<dnnl_graph_compiler_get_api_version_t>(
                        dlsym(dl, "dnnl_graph_compiler_get_api_version")),
                reinterpret_cast<dnnl_graph_compiler_create_t>(
                        dlsym(dl, "dnnl_graph_compiler_create")),
                reinterpret_cast<dnnl_graph_compiler_destroy_t>(
                        dlsym(dl, "dnnl_graph_compiler_destroy")),
                reinterpret_cast<dnnl_graph_compiler_compile_t>(
                        dlsym(dl, "dnnl_graph_compiler_compile")),
                reinterpret_cast<dnnl_graph_compiler_destroy_executable_t>(
                        dlsym(dl, "dnnl_graph_compiler_destroy_executable")),
                reinterpret_cast<dnnl_graph_compiler_execute_t>(
                        dlsym(dl, "dnnl_graph_compiler_execute")));
        return instance;
    }

private:
    dnnl_graph_compiler(
            dnnl_graph_compiler_get_api_version_t get_api_version,
            dnnl_graph_compiler_create_t create,
            dnnl_graph_compiler_destroy_t destroy,
            dnnl_graph_compiler_compile_t compile,
            dnnl_graph_compiler_destroy_executable_t destroy_executable,
            dnnl_graph_compiler_execute_t execute)
        : dnnl_graph_compiler_get_api_version(get_api_version)
        , dnnl_graph_compiler_create(create)
        , dnnl_graph_compiler_destroy(destroy)
        , dnnl_graph_compiler_compile(compile)
        , dnnl_graph_compiler_destroy_executable(destroy_executable)
        , dnnl_graph_compiler_execute(execute) {}
};
```

</details>

-----

And to make it portable:
<details>

```c++
#ifdef _WIN32
#include <windows.h>
#define DNNL_GC_LIB_NAME "graph_compiler.dll"
#define dlopen(libname, flags) LoadLibrary(libname)
#define dlsym(handle, funcname) GetProcAddress((HMODULE)handle, funcname)
#elif __APPLE__
#include <dlfcn.h>
#define DNNL_GC_LIB_NAME "libgraph_compiler.dylib"
#else
#include <dlfcn.h>
#define DNNL_GC_LIB_NAME "libgraph_compiler.so"
#endif
```

</details>

----

A basic (not complete) shared_ptr based implementation. The graph_compiler class has the compile() method, that returns a graph_compiler::executable - this is a wrapper around shared_ptr<graph_compiler> and dnnl_graph_compiler_executable*, it calls gc->destroy_executable(exe) in the destructor. The library loading/unloading is hidden from the user.

<details>

```c++
#ifdef _WIN32
#include <windows.h>
#define DNNL_GC_LIB_NAME "graph_compiler.dll"
#define dlopen(libname, flags) LoadLibrary(libname)
#define dlsym(handle, funcname) GetProcAddress((HMODULE)handle, funcname)
#define dlclose(handle) FreeLibrary((HMODULE)handle)
#elif __APPLE__
#include <dlfcn.h>
#define DNNL_GC_LIB_NAME "libgraph_compiler.dylib"
#else
#include <dlfcn.h>
#include <mutex>
#define DNNL_GC_LIB_NAME "libgraph_compiler.so"
#endif

struct graph_compiler {
  typedef dnnl_status_t (*dnnl_graph_compiler_get_api_version_t)(
      dnnl_graph_compiler_api_version *v);
  typedef dnnl_status_t (*dnnl_graph_compiler_create_t)(
      const struct dnnl_graph_compiler_context *ctx,
      const struct dnnl_graph_compiler **gc);
  typedef void (*dnnl_graph_compiler_destroy_t)(
      const struct dnnl_graph_compiler *gc);
  typedef dnnl_status_t (*dnnl_graph_compiler_compile_t)(
      const struct dnnl_graph_compiler *gc, const char *graph_json,
      const struct dnnl_graph_compiler_executable **exe);
  typedef void (*dnnl_graph_compiler_destroy_executable_t)(
      const struct dnnl_graph_compiler *gc,
      const struct dnnl_graph_compiler_executable *exe);
  typedef dnnl_status_t (*dnnl_graph_compiler_execute_t)(
      const struct dnnl_graph_compiler *gc,
      const struct dnnl_graph_compiler_executable *exe,
      dnnl_graph_compiler_tensor *inputs, dnnl_graph_compiler_tensor *outputs);

  struct executable {
    executable(const std::shared_ptr<const graph_compiler> &gc,
               const dnnl_graph_compiler_executable *exe)
        : gc(gc), exe(exe) {}
    ~executable() { gc->destroy_executable_fn(gc->gc, exe); }

    std::unique_ptr<executable> execute(dnnl_graph_compiler_tensor *inputs,
                                        dnnl_graph_compiler_tensor *outputs) {
      WRAP_GC_CALL(gc->execute_fn(gc->gc, exe, inputs, outputs),
                   "Execution failed");
      return std::make_unique<executable>(gc, exe);
    }

  private:
    std::shared_ptr<const graph_compiler> const gc;
    const dnnl_graph_compiler_executable *const exe;
  };

  graph_compiler(void *dl, const struct dnnl_graph_compiler *gc,
                 dnnl_graph_compiler_get_api_version_t get_version_fn,
                 dnnl_graph_compiler_destroy_t destroy_fn,
                 dnnl_graph_compiler_compile_t compile_fn,
                 dnnl_graph_compiler_destroy_executable_t destroy_executable_fn,
                 dnnl_graph_compiler_execute_t execute_fn)
      : dl(dl), gc(gc), get_version_fn(get_version_fn), destroy_fn(destroy_fn),
        compile_fn(compile_fn), destroy_executable_fn(destroy_executable_fn),
        execute_fn(execute_fn) {}
  ~graph_compiler() {
    destroy_fn(gc);
    dlclose(dl);
  }

  static std::shared_ptr<const graph_compiler> get() {
    static std::mutex mtx;
    static std::shared_ptr<const graph_compiler> instance;
    std::shared_ptr<const graph_compiler> ptr =
        std::atomic_load_explicit(&instance, std::memory_order_acquire);

    if (!ptr) {
      std::lock_guard<std::mutex> lock(mtx);
      ptr = instance;
      if (!ptr) {
        void *dl = dlopen(DNNL_GC_LIB_NAME, RTLD_LAZY);
        auto create = reinterpret_cast<dnnl_graph_compiler_create_t>(
            dlsym(dl, "dnnl_graph_compiler_create"));
        const struct dnnl_graph_compiler *gc;
        dnnl_graph_compiler_context ctx{0 /* ... */};
        WRAP_GC_CALL(create(&ctx, &gc), "Failed to create graph compiler");
        ptr = std::make_shared<graph_compiler>(
            dl, gc,
            reinterpret_cast<dnnl_graph_compiler_get_api_version_t>(
                dlsym(dl, "dnnl_graph_compiler_get_api_version")),
            reinterpret_cast<dnnl_graph_compiler_destroy_t>(
                dlsym(dl, "dnnl_graph_compiler_destroy")),
            reinterpret_cast<dnnl_graph_compiler_compile_t>(
                dlsym(dl, "dnnl_graph_compiler_compile")),
            reinterpret_cast<dnnl_graph_compiler_destroy_executable_t>(
                dlsym(dl, "dnnl_graph_compiler_destroy_executable")),
            reinterpret_cast<dnnl_graph_compiler_execute_t>(
                dlsym(dl, "dnnl_graph_compiler_execute")));
        std::atomic_store_explicit(&instance, ptr, std::memory_order_release);
      }
    }

    return ptr;
  }

  std::unique_ptr<executable> compile(const char *graph_json) {
    const dnnl_graph_compiler_executable *exe;
    WRAP_GC_CALL(compile_fn(gc, graph_json, &exe),
                 "Failed to compile the graph");
    return std::make_unique<executable>(get(), exe);
  }

private:
  void *const dl;
  const struct dnnl_graph_compiler *const gc;
  dnnl_graph_compiler_get_api_version_t const get_version_fn;
  dnnl_graph_compiler_destroy_t const destroy_fn;
  dnnl_graph_compiler_compile_t const compile_fn;
  dnnl_graph_compiler_destroy_executable_t const destroy_executable_fn;
  dnnl_graph_compiler_execute_t const execute_fn;
};
```

</details>

Reply:
_Originally posted by @dchigarev in https://github.com/intel-ai/oneDNN/pull/2#discussion_r1592724745_

> Instead of using mutex, I recommend to use either a singleton or shared_ptr.

Reworked to use singleton in this commit [cc9a6a6](https://github.com/intel-ai/oneDNN/pull/2/commits/cc9a6a6477d774d3c769537b76ef17e1de5cd079). The `graph_compiler_loader` now has only one public static method that returns a reference to a vtable.

> And to make it portable

Applied in this commit [b8c3ff5](https://github.com/intel-ai/oneDNN/pull/2/commits/b8c3ff546af7db78a7a4f897a298c64a70532557)

> The graph_compiler class has the compile() method, that returns a graph_compiler::executable - this is a wrapper around shared_ptr<graph_compiler> and dnnl_graph_compiler_executable*

In this implementation we mix responsibilities of a C++ wrapper over C-API and a library loader. I would like keep them separate. Anyway, I would like to introduce C++ wrappers (if needed) in a separate PR as this one is already quite big